bindings/pyroot/pythonizations/python/ROOT/_pythonization/_rdf_pyz.py

```diff
@@ -0,0 +1,277 @@
+        return self.func_call
+
+def _PyFilter(rdf, callable_or_str, *args , extra_args = {} ):
+    """
+    Filters the entries of RDF according to a given condition.
+    Arguments:
+        callable_or_str: The condition according to which the RDF is to be filtered. 
+            It can be either a python callable or a c style string.
+        *args:
+            Can be at most two of them.
+            a. The list of columns in  which case the starting arguments of the provided function callable in mapped to the columns in the list.
+                If not provide then it tries maps the name of the parameter to a column name of the RDF. Flags an error if not found.
+            b. The name of the Filter as a string.
+        extra_args: If the function needs any other parameters that are not columns and must be passed explicitly.
+    Returns:
+        self: The filtered instance of the RDF.
+    
+    Examples:
+    1. rdf.Filter("x>y")
+        Filter using a string containing c++ code 
+    2. rdf.Filter(lambda x,y: x>y)
+        Filter using a python callable.
+    3. rdf.Filter(lambda x,y: x>y, ["y", "x"])
+        Filter using a python callable with column list passed explicitly.
+        When the column list is passed, the function maps the starting input parameters to those columns instead.
+        Here the function param x is mapped to column y, and the second function param y is mapped to column x.
+        This is equivalent to the filter "y>x".
+    4. rdf.Filter(lambda col1, const1: col1>const1, ["x"], extra_args = {"const1":0.5})
+        Filter using a python callable with column list passed explicitly and extra_args.
+        Here the function parameter col1 is mapped to col "x" of the RDF. and the const1 parameter is mapped to the float value 0.5
+        This is equivalent to the filter x>0.5"
+
+    Example:
+    rdf = ROOT.RDataFrame(5).Define("x", "(double) rdfentry_")
+    def x_greater_than_2(x): 
+        return x>2
+    fj = FunctionJitter(rdf)
+    func_call = fj.jit(x_greater_than_2)
+    fil1 = rdf.Filter( "Numba::" + func_call, "x is greater than 2")
+
+    """
+    if isinstance(callable_or_str, str): # If string argument is passed. Invoke the Original Filters.
+        return rdf._OriginalFilter(callable_or_str, *args)
+
+    if not callable(callable_or_str): # The 1st argument is either a string or a python callable.
+        raise TypeError(f"The first argument of a Filter operation should be a callable. {type(callable_or_str).__name__} object is not callable.")
+    
+    if len(args) > 2: 
+        raise TypeError(f"Filter takes at most 3 positional arguments but {len(args) + 1} was given")
+    
+    # Check if it is a c++ callable.
+    import libcppyy
+    if type(callable_or_str) == libcppyy.CPPOverload:
+        return rdf._OriginalFilter(callable_or_str, *args)
+    if getattr(callable_or_str, 'target_type', None) == libcppyy.CPPOverload:
+        return rdf._OriginalFilter(callable_or_str(), *args)
```

Comment:
Why calling `callable_or_str()`? Also please add a comment here (in the line above the return, same indentation level) saying that this is an `std::function`.

Also, now I'm thinking what happens in the case of a function pointer?

Reply:
Ping :)

Reply:
I'll take out the ```callab_or_str()```
This was to account for the case in which the c++ function is wrapped by another function that returns the original function. Considering this was not allowed previously. NO point supporting it now.

So in all, if it is a cppyy callable, We pass it to the original filter any errors should be flagged at the C++ level hence.

Reply:
Wasn't the second case when the user provides a Python proxy of an std::function? This is supported now iirc, but that is not a callable that returns a callable, right?

Reply:
Indeed the std::function case is supported, it returns a proxy to std::function as expected and works within a filter. Its attribute `target_type` is indeed a `CPPOverload` so it should just be passed as is without calling it with parenthesis.
Example:
```python
import ROOT
ROOT.gInterpreter.Declare("""
bool filt(unsigned long long x) { return x > 5; }
""")

df = ROOT.RDataFrame(10).Define("x", "rdfentry_")
stdfn = ROOT.std.function["bool(unsigned long long)"](ROOT.filt)
print(df.Filter(stdfn, ["x"], "my filter").Count().GetValue())
print(f"{getattr(stdfn, 'target_type', None)=}")
```
Which outputs
```
$: python test_stdfn.py 
4
getattr(stdfn, 'target_type', None)=<cppyy.CPPOverload object at 0x7f49043dba30>
```

With this said, I then believe the correct approach here is
```suggestion
   # Second condition is a Python proxy to an std::function
   if (type(callable_or_str) is libcppyy.CPPOverload or
       type(getattr(callable_or_str, 'target_type', None)) is libcppyy.CPPOverload)  :
        return rdf._OriginalFilter(callable_or_str, *args)
```

Reply:
Yes, and it would be good to add a comment that clarifies what each condition of that if represents.

Reply:
Although consider that in general `CPPOverload` doesn't work with RDF
```
# print(df.Filter(ROOT.filt2, ["x"], "my filter").Count().GetValue())
# Errors with
# Traceback (most recent call last):
#   File "/home/vpadulan/projects/rootcode/pyz-filter-getfn/test_cppyy_fn_rdf.py", line 62, in <module>
#     print(df.Filter(ROOT.filt2, ["x"], "my filter").Count().GetValue())
# TypeError: Template method resolution failed:
#   ROOT::RDF::RInterface<ROOT::Detail::RDF::RJittedFilter,void> ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager,void>::Filter(basic_string_view<char,char_traits<char> > expression, basic_string_view<char,char_traits<char> > name = "") =>
#     TypeError: takes at most 2 arguments (3 given)
#   Failed to instantiate "Filter(CPPOverload,std::initializer_list<std::string>,std::string)"
```

Reply:
@vepadulano I think the problem in your last example is a bug with the initializer list, nothing to do with the callable. You can double check by passing an std::vector instead of a Python list with the column x

Reply:
It turns out there might be more that we're missing for this specific case, but it's outside of the scope of this PR